SSE with HTTP2 blogpost (#665)
ordina-jworks/ordina-jworks.github.io: _posts/2021-04-23-SSE-with-HTTP2.md

This means it does not require a special protocol like STOMP or MQTT which in turn requires server implementation to get it working making SSE a lot easier to set up.
SSE also has built in support for reestablishing connections and event IDs which WS lacks by design.

So the main question you have to ask yourself is whether your use-case fits into the whole SSE story.
Is there no communication from client to server required?
Is the application you are trying to build supposed to be cloud-native?

If the answer to these questions is no and there is no it might save you a lot of work going for SSE over WS.

# HTTP/1.1 vs HTTP/2
HTTP/1.1 is an old protocol, it loads requests one-by-one over a single TCP connection or in parallel over multiple TCP connections in an effort to decrease loading times while requiring more resources.
This was fine when this protocol was new, about 23 years ago, but as time goes by and webpages become more advanced, the limitations of this protocol are really starting to show.
This is why HTTP/2 was made, it aims to tackle the limitations set by HTTP/1.1 and be more future-proof.

> With HTTP/2 multiple requests can be sent over the same TCP connection with responses arriving out of order.
> HTTP/2 is a binary protocol, removing security issues and error-proneness that come with text-based protocols.
> It is backwards compatible with earlier versions of the protocol and is compatible with almost all browsers.
> HTTP/2 also avoids the round trip to the server by having the server intuitively sending resources that will be required to render the page.
> All these advantages eliminate the need for developers to write best practice workarounds to deal with the limitations of older versions of the protocol,
> they decrease loading times and improve the website infrastructure.
> This on top of full backwards compatibility make the choice between HTTP/1.1 and HTTP/2 for Server-sent events a no-brainer.

Reviewer: ```suggestion
With HTTP/2, multiple requests can be sent over the same TCP connection with responses arriving out of order.
HTTP/2 is a binary protocol, removing security issues and error-proneness that come with text-based protocols.
It is backward compatible with earlier versions of the protocol and is compatible with almost all browsers.
HTTP/2 also avoids the round trip to the server by having the server intuitively sending resources that will be required to render the page.
All these advantages eliminate the need for developers to write best practice workarounds to deal with the limitations of older versions of the protocol,
they decrease loading times and improve the website infrastructure.
This on top of full backward compatibility makes the choice between HTTP/1.1 and HTTP/2 for Server-sent events a no-brainer.
```

{:refdef: style="text-align: center;"}
<img src="{{ '/img/2021-04-23-SSE-HTTP2/HTTP1-vs-HTTP2.png' | prepend: site.baseurl }}" alt="Workspace" class="image" style="margin:0px auto; max-width:70%">
{: refdef}

# Summary

> The key takeaways from this post are that the choice between SSE and Websocket is entirely dependent on the use-case of the application you are trying to develop.
> If you are looking for a stateless approach, or you don't have a need for client to server communication, SSE might be the solution for you!
> The other takeaway is that you should definitely use HTTP/2 to get the most out of your application and not run into the limitations that HTTP/1.1 lays upon SSE.
> 
> If after reading this blogpost you have come to the conclusion that you would be better off building a Websocket application to fit your use-case you can read through a blogpost on websockets made by my colleague Kevin Van Houtte [here](https://ordina-jworks.github.io/event-driven/2020/06/30/user-feedback-websockets.html){:target="_blank" rel="noopener noreferrer"}.

Reviewer: ```suggestion
The key takeaways in choosing WS or SSE are entirely dependent on the use-case of the application you are trying to develop.
If you are looking for a stateless approach, or you don't have a need for client-to-server communication, SSE might be the solution for you!
The other takeaway is that you should definitely use HTTP/2 to get the most out of your application and not run into the limitations that HTTP/1.1 lays upon SSE.

If after reading this blogpost you have come to the conclusion that you would be better off building a Websocket application to fit your use-case you can read through a blog post on WebSockets made by my colleague Kevin Van Houtte [here](https://ordina-jworks.github.io/event-driven/2020/06/30/user-feedback-websockets.html){:target="_blank" rel="noopener noreferrer"}.
```

# Demo application
In this part of the blogpost I am going to show you how easy it is to develop your own SSE application.

### The use-case

> For the application we are going to build we are going to build a Spring Boot application that consumes a Chuck Norris joke REST API and then use a Flux to push joke data from the server using Server-Sent events to any clients that are subscribed.

Reviewer: ```suggestion
We are going to build a Spring Boot application that consumes a Chuck Norris joke REST API and uses Flux to push joke data from the server using Server-Sent events to any clients that are subscribed.
```
